src/scancode/resource.py

```diff
@@ -120,6 +120,39 @@ class ResourceNotInCache(Exception):
 class UnknownResource(Exception):
     pass
 
+def depth_walk(root_location, max_depth, skip_ignored=lambda x: False, error_handler=lambda:None):
+    """ Generator that walks the `root_location` directory recursively upto `max_depth` of 
+    nested subdirectories and yields a (top, dirs, files) tuple at each step. 
+
+    Arguments: 
+       - root_location: Absolute path for the directory to be walked
+       - max_depth: Positive integer for fixed depth limit. negative for no limit
+       - skip_ignored: Function that returns boolean indicating whether to ignore searching files
+                      in the current location. No ignoring by default
+       - error_handler: Error handler callback. No action taken by default.
+    """
+
+    # Find root directory depth using path separator's count
+    root_dir_depth = root_location.count(os.path.sep)
+
+    # Default behaviour. current_depth is always less than max_depth so that 
+    # os_walk keeps running. When max_depth is limited, the current_depth is
+    # changed in each iteration and compared with max_depth
+    current_depth = -2
+
+    for top, dirs, files in os_walk(root_location, topdown=True, onerror=error_handler):
+        # If depth is limited
+        if max_depth > 0:
+            current_depth = top.count(os.path.sep) - root_dir_depth
```

Comment:
Are you sure `count(os.path.sep)` will work OK? what if the path contains a segment with multiple contiguous separators?
for instance what happens with this:
```
>>> for x in os.walk('samples///'):
...  print(x)
... 
('samples///', ['arch', 'JGroups', 'zlib'], ['screenshot.png', 'README'])
('samples///arch', [], ['zlib.tar.gz'])
```

Reply:
The idea here is to normalize the `root_location` path first using `os.path.abspath` or `os.path.normpath`, so that any unnecessary slashes are removed. In our case, it being done in the `__init__` method of `Codebase`. That guarantees that any adjacent slashes are removed from the path.

Reply:
`depth_walk` itself does not guarantee that the path is normalized or absolute. Do you think we should have another `abspath` inside here?

Reply:
not needed. Your answer above makes sense :+1: